lib/srv/app/server.go

```diff
@@ -706,20 +707,27 @@ func (s *Server) authorize(ctx context.Context, r *http.Request) (*tlsca.Identit
 // getSession returns a request session used to proxy the request to the
 // target application. Always checks if the session is valid first and if so,
 // will return a cached session, otherwise will create one.
+// The in-flight request count is automatically incremented on the session.
+// The caller must call session.release() after finishing its use
 func (s *Server) getSession(ctx context.Context, identity *tlsca.Identity, app types.Application) (*sessionChunk, error) {
-	// If a cached forwarder exists, return it right away.
-	session, err := s.cache.get(identity.RouteToApp.SessionID)
-	if err == nil {
-		return session, nil
-	}
+	for {
```

Comment:
Is it really worth it to keep trying to fetch the session chunk from the cache (for an unlimited amount of times, even), if we're still potentially opening multiple session chunks for the same `SessionID` at the same time anyway?

This would read much cleaner if we just did this (together with having `newSessionChunk` return a refcount of 1):
```go
	session, err := s.cache.get(identity.RouteToApp.SessionID)
	if err == nil && session.acquire() == nil {
		return session, nil
	}

	session, err = s.newSessionChunk(ctx, identity, app)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	return session
```